### PR TITLE
[ASan][libc++] String annotations optimizations fix with lambda

### PR DESCRIPTION
**Pull request by AdvenamTacet**
This commit addresses optimization and instrumentation challenges encountered within comma constructors.
  1) _LIBCPP_STRING_INTERNAL_MEMORY_ACCESS does not work in comma constructors.
  2) Code inside comma constructors is not always correctly optimized. Problematic code examples:
        - `: __r_(((__str.__is_long() ? 0 : (__str.__annotate_delete(), 0)), std::move(__str.__r_))) {`
        - `: __r_(__r_([&](){ if(!__s.__is_long()) __s.__annotate_delete(); return std::move(__s.__r_);}())) {`

However, lambda with argument seems to be correctly optimized. This patch uses that fact.

Use of lambda based on idea from @ldionne.

### CODE REVIEW

**Comment by AdvenamTacet** [comment]
This does not answer problem of `_LIBCPP_STRING_INTERNAL_MEMORY_ACCESS` not working inside comma constructors and therefore we have additional write to shadow memory while using ASan. That's a small cost, which I think we can ignore.

Alternative is a function like that:
```cpp
  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_STRING_INTERNAL_MEMORY_ACCESS
  __compressed_pair<__rep, allocator_type> && __compressed_pair_id(basic_string &__str) {
    if (!__str.__is_long())
      __str.__annotate_delete();
    return std::move(__str.__r_);
  }

```

But I support lambda solution.

File: libcxx/include/string [hunk 1]
```diff
@@ -922,7 +922,10 @@ public:
       // Turning off ASan instrumentation for variable initialization with _LIBCPP_STRING_INTERNAL_MEMORY_ACCESS
       // does not work consistently during initialization of __r_, so we instead unpoison __str's memory manually first.
       // __str's memory needs to be unpoisoned only in the case where it's a short string.
-      : __r_(((__str.__is_long() ? 0 : (__str.__annotate_delete(), 0)), std::move(__str.__r_))) {
+      // Lambda is used because of optimization challenges encountered within comma constructors.
+      // Lambda with argument is correctly optimized, but it does not solve the problem with internal memory
+      // access macro.
```
Comment (ldionne) [inline libcxx/include/string:927]: IMO this comment doesn't add information we want to retain in the code, and I would remove it. I think this kind of information is best found in the commit message.

**Review by ldionne — APPROVED** [review]
I'm fine with this (with a nitpick about the comment, which I think doesn't add much). This is kind of just a cleaner way to write the same thing as before, so even better if it fixes an optimization problem.

If this were making the code harder to read or more complicated, I'd ask for at least a regression test but in this case we can pretend that it's just a slight "refactoring", even though it is motivated by fixing a regression. It would be good to have some kind of regression test is that's feasible, though.

**Comment by ldionne** [comment]
Since this change is motivated by issues reported by @EricWF, I think it would be relevant for Eric to stamp this change as well, to make sure it solves at least some of the problems he was seeing.

**Comment by AdvenamTacet** [comment]
It also makes the code easier to read. Code inside of the lambda is easy to read, instead of being a hack with comma operator.

**Review by EricWF — CHANGES_REQUESTED** [review]
LGTM, but I would like to see some actual assembly before and after in the review thread. Could you provide a godbolt or diff of the assembly generated?

I found some other things unrelated to this review while looking at this, which I'll bring up offline.

**Comment by AdvenamTacet** [comment]
@EricWF the difference is the same as described here: https://github.com/llvm/llvm-project/pull/76192#issuecomment-1867058465
Basically asm goes back to the state from before any string annotations on know to me examples.

I can create dedicated examples to that patch on Monday.

I also see a big benefit in readability.

**Review by EricWF — APPROVED** [review]
Actually I'm happy to generate the examples myself.

You've won me on the readability argument alone.

**Comment by AdvenamTacet** [comment]
It looks like this PR made a buildbot fail.
https://lab.llvm.org/buildbot/#/builders/37/builds/29650

I'm trying to understand why.

**Pull request merged**